File: 0198-house-robber/0198-house-robber.cpp
class Solution {
public:
    int rob(vector<int>& nums) {
        int n = nums.size();

        if (n == 1) {
            return nums[0];
        }
        vector<int> earns(n, 0);

        earns[0] = nums[0];
        earns[1] = max(nums[0], nums[1]);

        for (int i = 2; i < n; i++) {
            earns[i] = max(earns[i - 1], earns[i - 2] + nums[i]);
        }

        return earns[n - 1];
    }
};

File: 0198-house-robber/0198-house-robber.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        # if len(nums) == 1:
        #     return nums[0]
        # x = nums[0]
        # y = max(nums[0], nums[1])

        # for i in range(2, len(nums)):
        #     z = y
        #     y = max(x + nums[i], y)
        #     x = z

        # return max(x, y) 
        
        if len(nums) == 1:
            return nums[0]

        x = nums[0]
        y = max(nums[0], nums[1])

        for i in range(2, len(nums)):
            z = y
            y = max(y, x + nums[i])
            x = z
        
        return y